mountpoint-s3/src/upload.rs

```diff
@@ -136,9 +145,19 @@ impl<Client: ObjectClient> UploadRequest<Client> {
     pub async fn complete(self) -> Result<PutObjectResult, PutRequestError<Client>> {
         let size = self.size();
         let checksum = self.hasher.finalize();
-        self.request
+        let result = self
+            .request
             .review_and_complete(move |review| verify_checksums(review, size, checksum))
-            .await
+            .await?;
+        self.sse
+            .verify_response(result.sse_type.as_deref(), result.sse_kms_key_id.as_deref())
+            .unwrap_or_else(|err| {
+                panic!(
```

Comment:
I've found surprising that panicing here does not result in Mountpoint being stopped (the same applies for `panic!` on line 95 in `UploadRequest::new`), it just prints the stacktrace and logs the failure for a single thread:
```
thread 'fuse-worker-3' panicked at /local/home/vlaad/local/mountpoint-s3/mountpoint-s3/src/upload.rs:155:17:
SSE settings were corrupted, object c0cfb6fe22bc40c8b0c8c35ecbb931c8 WAS uploaded to S3, panicing, error: SSE type mismatch. expected: Some("aws:kms"), actual: Some("aws:kms")
2024-02-14T16:50:27.421974Z  WARN fuser::reply: Reply not sent for operation 50, replying with I/O error
```

user application fails on `flush()` as expected, here is the output of a python script:
```
Traceback (most recent call last):
  File "/local/home/vlaad/local/mountpoint-s3/writer.py", line 24, in <module>
    f.flush()
OSError: [Errno 5] Input/output error
```

The behaviour which we were discussing previously (and I assumed it here) is [aborting on panic](https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html). I believe we don't want to enable this for all `panic!`'s in our code. 

Keeping the status quo of this PR, when a thread panicked while the rest of the system keeps working, seems like not a good choice (e.g. is the panicked thread even restarted afterwards?). I think, we need to return an error instead of panicking (will likely require changing signature at least in `UploadRequest::new` case). And then 2 options:
1. accept that we keep MP running in those cases (as you can see `flush` returns an error to the user in any case, so this is a good signal that things went wrong);
2. work on a mechanism that gracefully stops Mountpoint on certain errors (somewhere at [fs.rs](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/fs.rs#L243) level?)

I'd go with 1/, what do you think is the best here?

Reply:
by changing signature of `UploadRequest::new` I mean something like this:
https://github.com/vladem/mountpoint-s3/pull/5/files

Reply:
`flush` returns an error, and that makes it to the user as a return code from `close(fd)`, but people rarely check that return code.

The other option is `std::os::exit(1)` if you want to abort the process.

Reply:
Hm, I see the point about unreliability of `flush` in error signalling. What I do not like about `exit()` is that it doesn't do any clean up. For example if there were simultaneous MPU uploads, they may be [left](https://aws.amazon.com/blogs/aws-cloud-financial-management/discovering-and-deleting-incomplete-multipart-uploads-to-lower-amazon-s3-costs/) incomplete on S3 side.

I think we still don't want to panic on `UploadRequest::new` as returning an error from there will also make `write()` to [return](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/fs.rs#L137) an `EIO`/`EPERM`, which I expect most users to check. And no data was uploaded yet, so its not a high risk to overlook the error.

For a "nuclear" case of `UploadRequest::complete` (when the data is known to be uploaded already and the corruption happened in transit or on S3 side) I think it's acceptable to do `std::os::exit(1)`.

Reply:
Implemented error handling as described in the last comment in a separate branch. Can merge it to this one if we agree its better than `panic`ing: https://github.com/vladem/mountpoint-s3/pull/5

Reply:
Discussed with @passaro, and decided to exit in both cases: on `UploadRequest::new` and on `UploadRequest::complete`. In a follow-up PR I am going to change `UploadRequest::new` to return an error, in a similar way as it is done here: https://github.com/vladem/mountpoint-s3/pull/5